### PR TITLE
aur-search: use intersection of results for multiple terms

### PR DESCRIPTION
**Pull request by maximbaz**
This is a prototype for #313.

Now the behavior of `aur search` matches `pacman -Ss`, when multiple search terms are provided, only packages that match ALL of the terms are returned.

I didn't create an option to keep the old behavior because I'm not sure if it is actually needed.

Feedback is welcome 🙂

image

### CODE REVIEW

**Comment by AladW** [comment]
I'm working on a "fold" function for aur search, so that with long arrays like make/depends the output is properly wrapped like with `cower`/`pacman`. Once I'm done with that I'll take a look. :+1:

**Comment by maximbaz** [comment]
Rebased to include new fields License and Keywords, as well as renamed "First Submitted" to "Submitted". Hope I didn't miss anything else.

image

**Comment by maximbaz** [comment]
Although to be fair I don't see the License and Keywords because they are not returned by aur-rpc...

image

**Comment by AladW** [comment]
Yeah, Keywords and License are not available in the search output. (Hence why I set an array with a (null) field in case those fields are absent, so json_long works with both search and info output)

**Comment by maximbaz** [comment]
Aaaah... Then I'm happy to report that I did not break that functionality either 😄 

image

**Comment by Morganamilo** [comment]
This still queries the RPC for each term right? If you only want the results that match all the terms then can't you just search the RPC for the first term. Then filter the results against the other terms. That should be that same and take 1 RPC request instead of N.

**Comment by maximbaz** [comment]
It is possible, but I want to keep the change simple and not reimplement the filtering, don't want to bother with accounting for the field that it is being searched by, etc. Also, it sounds scary when you say N calls, but in reality 99.9% of usages will issue not more than... 3 RPC calls, I would say 🙂

**Comment by AladW** [comment]
Actually, it shouldn't be complicated thanks to the power of `jq`...

https://rosettacode.org/wiki/Set#Finite_Sets_of_JSON_Entities

**Comment by Morganamilo** [comment]
Yeah it might be a bit of over optimization but I do like to minimize the use of the RPC.

**Comment by maximbaz** [comment]
Oh, and I thought `bash` syntax is weird, but now I've seen some `jq`...

```
# If A and B are sets, then intersection(A;B) emits their intersection:
def intersection(A;B):
  (A|length) as $al
  | (B|length) as $bl
  | if $al == 0 or $bl == 0 then []
    else 
      reduce range(0; $al + $bl) as $k
        ( [0, 0, []];
          .[0] as $i | .[1] as $j
          | if $i < $al and $j < $bl then
              if A[$i] == B[$j] then [ $i+1 , $j+1,  .[2] + [A[$i]]]
              elif  A[$i] < B[$j] then [ $i+1 , $j,  .[2] ]
              else [ $i , $j+1,  .[2] ]
              end
            else .
            end
         ) | .[2]
    end ;
```

**Comment by maximbaz** [comment]
By looking a bit more I don't feel comfortable going this way, how about merging the current approach and improving in a separate iteration if needed? The split between parsing and printing in this PR will be useful for every approach.

**Comment by AladW** [comment]
There seem to be some unrelated changes in the commits so it'll take some time to review

File: lib/aur-search [hunk 1]
```diff
@@ -9,63 +9,51 @@ query_type=search
 search_by=name-desc
 sort_key=Name
 
-json_short() {
-    local Name Version NumVotes Maintainer OutOfDate Description
+parse_short() {
```
Comment (AladW) [inline lib/aur-search]: unrelated name change

File: lib/aur-search [hunk 2]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
+    jq -r --arg key "$1" '[.results[]] | sort_by(.[$key])[] | [
+        .Name         // "-",
```
Comment (AladW) [inline lib/aur-search]: unrelated change (and `//` only makes sense on fields that might be null or empty, never the case for name, version e.a.)

File: lib/aur-search [hunk 3]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
+        .Version      // "-",
+        .NumVotes     // "-",
+        .Maintainer   // "-",
+        .OutOfDate    // "-",
+        .Description  // "-"
+        ] | @tsv'
```
Comment (AladW) [inline lib/aur-search]: unrelated change, no benefit from going tsv instead of newline separated (more compact notation but makes tools like awk and read that use newlines by default harder to use)

File: lib/aur-search [hunk 4]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
-        read -r Name
-        read -r Version
-        read -r NumVotes
-        read -r Maintainer
-        read -r OutOfDate
-        read -r Description
-    }; do
-        [[ "$OutOfDate"  ]] && OutOfDate="(Out-of-date: $(date -d @"$OutOfDate" '+%d %B %Y'))"
-        [[ "$Maintainer" ]] && unset Maintainer
+output_short() {
```
Comment (AladW) [inline lib/aur-search]: unrelated name change

File: lib/aur-search [hunk 5]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
+    local Name Version NumVotes Maintainer OutOfDate Description
 
+    while IFS=$'\t' read -r Name Version NumVotes Maintainer OutOfDate Description ; do
+        [[ "$OutOfDate"  == "-" ]] && OutOfDate=""  || OutOfDate="(Out-of-date: $(date -d @"$OutOfDate" '+%d %B %Y'))"
```
Comment (AladW) [inline lib/aur-search]: unrelated change (old checkout?)

File: lib/aur-search [hunk 6]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
+        [[ "$Maintainer" != "-" ]] && Maintainer="" || Maintainer="(Orphaned)"
         printf "${BLUE}aur/${ALL_OFF}${BOLD}%s ${GREEN}%s ${ALL_OFF}(%s) ${RED}%s %s${ALL_OFF}\\n    %s\\n" \
-               "$Name" "$Version" "$NumVotes" "${Maintainer+(Orphaned)}" "$OutOfDate" "$Description"
+               "$Name" "$Version" "$NumVotes" "$Maintainer" "$OutOfDate" "$Description"
```
Comment (AladW) [inline lib/aur-search]: as L27

File: lib/aur-search [hunk 7]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
     done
 }
 
-json_long() {
+parse_long() {
```
Comment (AladW) [inline lib/aur-search]: As L14

File: lib/aur-search [hunk 8]
```diff
@@ -9,63 +9,51 @@ query_type=search
...
+    jq -r --arg key "$1" '[.results[]] | sort_by(.[$key])[] | [
```
Comment (AladW) [inline lib/aur-search]: As L20

File: lib/aur-search [hunk 9]
```diff
@@ -121,13 +109,43 @@ case $query_type in
     search) format=${format-short} ;;
 esac
 
-# set filters
+# set parsers
```
Comment (AladW) [inline lib/aur-search]: unrelated change, see also https://en.wikipedia.org/wiki/Filter_(software)

**Comment by maximbaz** [comment]
I'll leave some comments to explain the changes and hopefully make review easier:

1. Functions `json_short` and `json_long` are split in two stages - parsing PRC results and printing. 

1. Because parsing is separate now, previously the output of `jq` parsing was immediately read, now it is converted via `@tsv` and being read later from this new format - this introduced problem with fields that are missing a value, `read` cannot handle `\t\t`. That's why I made `-` to be a default value for every single field that `jq` is trying to parse, to avoid any potential for future issues with `read`.

1. Because missing value is marked as `-` instead of empty string, detecting `OutOfDate` and `Orphaned` changed a bit too.

1. `Keywords` and `License` changed their default value from `(null)` to `-` for consistency with the rest of the fields.

1. Intersection of results is done with grep and temporary files, most of that code was inspired by similar pieces in `aur-sync`.

File: lib/aur-search [hunk 10]
```diff
@@ -121,13 +109,43 @@ case $query_type in
...
+case $format in
+     long) parse() { parse_long  "$sort_key"; } ;;
+    short) parse() { parse_short "$sort_key"; } ;;
+esac
+
+# set output
 case $format in
-     long) parse() { json_long  "$sort_key"; } ;;
-    short) parse() { json_short "$sort_key"; } ;;
+     long) output() { output_long;  } ;;
```
Comment (AladW) [inline lib/aur-search]: As L12

File: lib/aur-search [hunk 11]
```diff
@@ -121,13 +109,43 @@ case $query_type in
...
+    short) output() { output_short; } ;;
```
Comment (AladW) [inline lib/aur-search]: As L14

File: lib/aur-search [hunk 12]
```diff
@@ -121,13 +109,43 @@ case $query_type in
...
+source /usr/share/makepkg/util/util.sh
+cd_safe "$tmp"
+
 # pipeline
-printf -- '%s\n' "$@" | aur rpc -t "$query_type" -b "$search_by" | parse
+for query in "$@"; do
+    printf -- '%s\n' "$query" | aur rpc -t "$query_type" -b "$search_by" | parse > query_results
+    if [[ -f query_results_combined ]]; then
+        mv query_results_combined query_results_prev
+        grep -Fxf query_results query_results_prev > query_results_combined
```
Comment (AladW) [inline lib/aur-search]: Does this preserve the validity of the JSON arrays? E.g. things like resultcount will definitely be wrong

**Comment by AladW** [comment]
I started the review before viewing your comment. I can't say I'm very fond of all these changes, since they affect large parts of that code without this being actually required for intersection functionality.

File: lib/aur-search (hunk 12 above)
Comment (AladW) [reply lib/aur-search]: Or I guess it parses the results first to TSV format, and then runs grep on it line-wise? Seems problematic if you want to expand functionality later

**Comment by maximbaz** [comment]
Well yes, changes look large, but do you see any actual issues with them? Splitting monolithic "parse-and-print" into two components sounds like a good idea anyway, having `-` as a default value for all fields is just for consistency and possible issues with RPC when suddenly some fields will become missing.

File: lib/aur-search (hunk 3 above)
Comment (AladW) [reply lib/aur-search]: https://github.com/AladW/aurutils/pull/328#discussion_r176923571

**Comment by AladW** [comment]
Well, with the old code you would immediately see which fields would be potentially null/empty in the RPC output, now you can't. I think that's more valuable than a potential scenario where suddenly the AUR starts behaving in erratic ways (haven't seen that in 2 years anyway)

File: lib/aur-search (hunk 12 above)
Comment (maximbaz) [reply lib/aur-search]: Exactly - why is it problematic though? When a more fancier functionality is available, it can read the TSV values in variables and do whatever it needs to do, it will have to split the current "parse-and-print" function anyway.

**Comment by maximbaz** [comment]
I don't mind removing the default `-` from the fields that cannot ever be empty 🙂

File: lib/aur-search (hunk 2 above)
Comment (maximbaz) [reply lib/aur-search]: Is it documented somewhere which exactly fields are optional and which are required? I was hoping it would be documented [here](https://wiki.archlinux.org/index.php/AurJson), but it doesn't seem to be the case.

Comment (AladW) [reply lib/aur-search]: I don't think so, unless you go investigate the PHP code! :grimacing:  What was there before is purely empirical

Comment (maximbaz) [reply lib/aur-search]: Then I see even less reasons to remove the default that saves us from potentially missing values... But I pushed this nevertheless, used the empirical data from `master` 🙂

**Comment by AladW** [comment]
All the other things in the review need fixing too.

Note that the whole grep/file approach should not be required since you can just use `uniq -d`.
```
printf -- '%s\n' "$@" | aur rpc ... | jq ... @tsv | sort | uniq -d | while read ...
```

File: lib/aur-search (hunk 1 above)
Comment (maximbaz) [reply lib/aur-search]: Functions `json_short` and `json_long` are split in two stages - parsing PRC results and printing. This is parsing, thus named `parse.

File: lib/aur-search (hunk 2 above)
Comment (maximbaz) [reply lib/aur-search]: This is done

File: lib/aur-search (hunk 3 above)
Comment (maximbaz) [reply lib/aur-search]: This is needed for removing duplicates.

File: lib/aur-search (hunk 4 above)
Comment (maximbaz) [reply lib/aur-search]: Functions `json_short` and `json_long` are split in two stages - parsing PRC results and printing. This is the printing the output part, thus named `output`.

File: lib/aur-search (hunk 5 above)
Comment (maximbaz) [reply lib/aur-search]: Because missing value is marked as `-` instead of empty string, detecting `OutOfDate` and `Orphaned` changed a bit too.

File: lib/aur-search (hunk 6 above)
Comment (maximbaz) [reply lib/aur-search]: Replied there

File: lib/aur-search (hunk 7 above)
Comment (maximbaz) [reply lib/aur-search]: Functions `json_short` and `json_long` are split in two stages - parsing PRC results and printing. This is parsing, thus named `parse`.

File: lib/aur-search (hunk 8 above)
Comment (maximbaz) [reply lib/aur-search]: Removed default values from the field, this should be done.

File: lib/aur-search (hunk 9 above)
Comment (maximbaz) [reply lib/aur-search]: if the functions are called `parse` and they are doing parsing, I think the comment should also say that

File: lib/aur-search (hunk 10 above)
Comment (maximbaz) [reply lib/aur-search]: answered there

File: lib/aur-search (hunk 11 above)
Comment (maximbaz) [reply lib/aur-search]: answered there

**Comment by maximbaz** [comment]
Could you please highlight which exactly things need fixing? I replied to every comment.

Thanks for the `uniq -d` hint, will check it out!

**Comment by AladW** [comment]
The changes are still to invasive for me to merge in a single PR. I'll probably give a shot at it myself with your `@tsv` idea and see what turns up

**Comment by maximbaz** [comment]
Okay! 👍

**Comment by AladW** [comment]
This is what I have in mind:
* call the while/read functions `fmt_long` and `fmt_short`, in line with `aur-fetch`, or `parse_long` and `parse_short` in line with the used pipeline
* call the `jq` functions `tabulate_long` and `tabulate_short`, to emphasize on `@tsv` and in line with `aur-deps-rpc`
* use `read -a` in `fmt_long` to save some space and since the fields are already clear from the `printf` lines
* leave out changes in format (hypens et al) for a separate commit or PR
* add an `-a` option which returns the union (`uniq -u`) instead of the intersection (`uniq -d`) of matched packages.

The end pipeline should look as follows:
```
printf -- '%s\n' "$'" | aur rpc -t "$query_type" -b "$search_by" | tabulate "$sort_key" | combine | parse
```

**Comment by maximbaz** [comment]
Sounds very good! Just make sure that `read` doesn't break on tabulated string that is missing some values, that was the whole point of me adding `-` and changing formats (so that instead of `\t\t` I have `\t-\t` to parse) - maybe there is a way to achieve the same without changing the format that I didn't find 🙂

**Comment by AladW** [comment]
https://lists.gnu.org/archive/html/help-bash/2016-05/msg00041.html

Any drawback in using `@csv` instead of `@tsv`?
```
user@v22016103960739491:~$ string=a,b,,d
user@v22016103960739491:~$ IFS=, read -a array <<< "$string"
user@v22016103960739491:~$ echo ${array[@]}
a b d
user@v22016103960739491:~$ echo ${array[0]}
a
user@v22016103960739491:~$ echo ${array[1]}
b
user@v22016103960739491:~$ echo ${array[2]}

user@v22016103960739491:~$ echo ${array[3]}
d
```

**Comment by maximbaz** [comment]
Don't think so, I'm sure jq is smart enough to handle possible commas in the received data itself, but worth checking anyway.

**Comment by AladW** [comment]
`jq` does the right thing but `read` has no knowledge about quoting:
```
user@v22016103960739491:~$ tabulate_short Name < out
"python2-octaviaclient","1.4.0-1",0,"andybz","","OpenStack Client plugin for Octavia, an OpenStack Load Balancing project"
user@v22016103960739491:~$ echo ${foo[5]}
"OpenStack Client plugin for Octavia
user@v22016103960739491:~$ echo ${foo[6]}
an OpenStack Load Balancing project"
```

**Comment by AladW** [comment]
At the risk of angering the bash gods, we could use `@sh` and `eval`:

https://github.com/stedolan/jq/wiki/Cookbook#extract-data-and-set-shell-variables

**Comment by maximbaz** [comment]
In my opinion `@tsv` and default value `-` is better, but perhaps I just don't see something bad about this approach that you do.

**Comment by AladW** [comment]
Yeah, I can't think of anything better either (the alternatives above are hardly satisfactory). I do something similar in `aur-vercmp` for `join(1)` use anyway.

**Comment by AladW** [comment]
Actually we don't need two jq functions, one suffices when you can ignore what's not required:
```
while IFS=$'\t' read -r Name _ Version Description _ _ _ NumVotes _ Maintainer OutOfDate _ _; do
```

**Comment by maximbaz** [comment]
Nice finding!

**Comment by AladW** [comment]
I ended up rewriting most of the script: https://github.com/AladW/aurutils/commit/c07e9571bdb4f7b0644b2700ac6069bcfc56df2e

Count me surprised if it still works.

**Comment by AladW** [comment]
1. uniq -d prints all duplicate lines, if there are none to begin (in the case of one argument) with it prints none
2. I forgot to set the separator to tab for column(1), should be `column -s $'\t'`
3. on info searches, the union of results is already done by the AUR (and the intersection makes no sense to begin with), so it should be disabled in that case.

I've also changed the code so that column is only run once for every search instead of once for every result. An issue that remains is that `column` ignores empty lines. On Debian there's a `column -e` option, but in the actual non-patched-to-death version `-e` seems to mean something else entirely...

**Comment by maximbaz** [comment]
1. I see, so basically we need to skip the `combine` step if there is only one entry in the `$@`?
2. Yep, that fixes the issue except for the missing empty lines. If there is no workaround, we can still run column once per result, and print `\n` after that, doesn't sound too bad...
3. So `combine` step should not only be skipped if there is only one entry in `$@`, but also if `-i` is provided?

**Comment by AladW** [comment]
Running column once per result is fine if there are few results, but will slow things down drastically if there are a few thousands...

I've used `sed` for now but I'm hoping to find some better solution. All in all, while I think the move to `@tsv` is generally useful, I think the union and intersection of `search` results should be handled on the AUR side. Compare the ongoing discussion on aur-dev:

https://lists.archlinux.org/pipermail/aur-dev/2018-March/004495.html

**Comment by AladW** [comment]
dcab37a added some reminders to the code, so I think this can be closed. It would still be a good idea to engage in the above aur-dev discussion though.

**Comment by maximbaz** [comment]
I confirm that all the issues are fixed, thanks 🙂

Agree that it should be implemented on the AUR side.

> \# XXX add option for intersection or only leave -a?

I'd say intersection should stay default and without an option, because that's how `pacman -Ss` works. Consistency 🙂

**Comment by AladW** [comment]
Works for me. Now we just need a man page update...

**Comment by AladW** [comment]
One issue remains, and that is whatever I do, I can't use both `join(" ")` and `//` for Keywords/License. The output is always empty.

As powerful as jq is, I swear malbolge has a more pleasant UI at times...
